### ISSUE TITLE
FarmOS Offline Mode

### ISSUE THREAD

**Issue by DomtronVox** [farmOS/farmOS#37]
Hello! First, thank you for creating and maintaining this project.

I was wondering if it would be possible to have an "offline mode" for FarmOS. That is, allow FarmOS to work without an internet connection.

For example my family lives out where there is no real internet connection to speak of. However, I have a local network that covers the house, which I plan to extend out to the main fields as well, with a personal web server. I have FarmOS running and can access it with all local computers and tablets, but each page load is very slow and takes up the precious little bandwidth I do have because it has to fetch external fonts, stylesheets, and scripts.

I would like to use FarmOS to track information, but the requirement for an internet connection makes FarmOS almost useless to me. :(  I'm mostly looking for instructions on how to remove**, or better yet host locally**, all the parts requiring an internet connection. Though it would be really nice if it was an option/step in the installation process. 

Thank You!

**Comment by mstenta** [comment]
Hi @DomtronVox - great question! This is something I've thought a lot about, and we actually discussed it on the last monthly call this past Wednesday.

The main things that depend on an internet connection in farmOS (off the top of my head) are:

1. Google Maps - farmOS uses Google Maps layers for it's default maps. However, it is possible to turn off the maps entirely in farmOS. The setting for that is at: /admin/config/farm/map
2. Openlayers library - By default, farmOS is set up to load the Openlayers javascript library via a CDN. This only happens if the maps are turned on, so the setting described above is also a workaround for this. Alternatively, it is possible to set up the Openlayers Drupal module to load the library from local files, rather than from a CDN.
3. Bootstrap library - The Drupal theme for farmOS (farm_theme) is based on the Bootstrap theme (http://drupal.org/project/bootstrap), and similarly it also loads the Bootstrap library JS and CSS from a CDN by default. It is possible to host those files locally, as well.

So ultimately, right now it is mainly the maps that need an internet connection - for the Google layers. If you turn off the maps, then that covers points 1 and 2 above. If you set up the Bootstrap CSS and JS files locally, then that will cover point 3.

(You also mentioned fonts... any custom font requirements must be coming from the Bootstrap library - so I would assume they would also be included with a local install of the Bootstrap library.)

I wish I could provide more detailed instructions for setting up Openlayers and/or Bootstrap locally, but I haven't done it in a while. I think both projects provide documentation for that, though:

https://www.drupal.org/project/bootstrap

https://www.drupal.org/project/openlayers

The one thing to be aware of is the version numbers.  I think we're using the latest version of Bootstrap, but we're actually using an older version of Openlayers because there were some compatibility issue with the later releases that still need to be fixed in the Drupal module.

I have given thought to packaging the libraries in farmOS by default - instead of using the CDN. I think that's still open for consideration. Perhaps we should create a feature request specifically for that... and maybe that could also include an option during install to use CDNs instead.

In other, related, news... @alexadamsmith is planning to build a simplified native android/iphone app for farmOS - which would allow some things to be done without an internet connection, and then sync back to farmOS when internet is available. Here is the current discussion around that: https://github.com/farmOS/farmOS-mobile-app/issues/2

**Comment by mstenta** [comment]
I'll also throw this into the mix: it is theoretically possible to host your own map layers and load them into your farmOS via Openlayers... ie: if you create your own aerial imagery with a drone or weather balloon. I've never tried this, so I'm not familiar with the process, but I'm pretty sure the Openlayers module support it.

So if you really wanted maps, but still wanted it to work without internet, you may be able to host the Openlayers library AND the map layers themselves locally

**Comment by mstenta** [comment]
Clarification on my last comment: I actually HAVE worked with custom layers, but they are hosted on MapKnitter.org - so that still requires an internet connection. I have not tried hosting map imagery myself. I know it's possible to set up your own map hosting server, too... ie with http://geoserver.org/

**Comment by tormi** [comment]
@DomtronVox , do you mean offline Drupal instance in your laptop or [website you can use offline](http://offlinefirst.org/) and sync back when online?

**Comment by mstenta** [comment]
@tormi and @DomtronVox - In the long term, I hope to see farmOS capable of working offline natively - using HTML5 and localStorage. That's the other thing we talked about on this month's call. Essentially: a one-page UI that you load on your phone when you have internet, which stores itself (and some data) in localstorage so it can continue to work offline - and then sync back up with the server when a connection is available. I think that's what you're referring to @tormi ... right?

Upgrading farmOS to Drupal 8 will be a good step towards that. There has been a lot of work put towards API-first architecture and progressive decoupling in 8 - making it a lot easier and more standardized to create HTML+JS UIs that access the backend via API but are not directly tied to it.

Until then, a native mobile app for android/iphone will be a great help for simple things like observations and checklists.

**Comment by tormi** [comment]
>  I think that's what you're referring to

@mstenta, yes

**Comment by DomtronVox** [comment]
@mstenta Thanks so much for the great info! Much more then I was hoping for. I just assumed there would be no way to get the map working offline and was focused on fixing your #3 list item so it works offline. 

@tormi I was talking about a local druple server that worked without the internet, not a web page app that can resync. Thanks for the response!

I'll look into installing bootstrap locally once I get some free time, and maybe put together some documentation after the fact. But it would defiantly be nice to have an install option for turning on/off CDN sources.

Sorry this response was made much later. I was expecting github to email me any responses, but it didn't send me any. :/ 

Again thanks for the info!

**Comment by jgaehring** [comment]
@mstenta Do you still have any interest in developing an offline-first web front end? I've been looking for some OSS projects to contribute to, particularly projects related to agriculture, and would be happy to pitch in. I've done some light work with service workers and PWA's in the past, but would love to tackle a rich problem like this. I'm primarily a JS/React developer, but think I could get up to speed pretty quickly on the Drupal environment, especially if you're migrating to a more API-oriented structure. Let me know if you'd be open to contributions like this in the future. Cheers!

**Comment by mstenta** [comment]
@jgaehring Yes! We would love help with this!

I think for it to be done effectively, though, there are some big dependencies we need to finish first. The biggest is the upgrade to Drupal 8: https://www.drupal.org/node/2361735

D8 (and the Drupal community in general) is shifting very heavily towards API first development, and I have been excited to steer the farmOS ship in that direction. D8 comes with a set of API modules specifically for enabling rich JS applications to be built on top and serve as the frontend (instead of Drupal's default frontend).

Some of this capability is available in Drupal 7, and we may be able to start right away, but it will have to be done in a way that plans for the D8 upgrade, so it doesn't end up holding up that process even more.

The other consideration is that farmOS (and Drupal) provide a LOT of UI - and it will take a lot of work to replicate that in a frontend JS application.  So I think we should start small, and add pieces iteratively over time.  It would be great if we could start with a simple "Offline" toggle in default (Drupal) UI that dynamically switches you over to the JS UI.  So users would want to take their farmOS out into the field could switch it to offline mode (which would sync the necessary data into their local storage), and then when they are back in range of wifi they could toggle back to get the full farmOS UI again.  Then, over time, more and more of the "full" UI can be replicated in the JS UI - until eventually that can take over as the default, and works both online and off.

That's my long-term vision.

But, I am not a Javascript developer, so I will definitely be looking for help with this!  My goal is to get farmOS upgraded to Drupal 8 ASAP so that it makes the process easier and opens farmOS up to the world of JS developers more.

Notably, with regards to React, there seems to be a lot of discussion around possibly adopting that framework in Drupal core in the future... so that's something to keep an eye on as well! https://www.drupal.org/blog/drupal-looking-to-adopt-react

**Comment by jgaehring** [comment]
Ah yes, I sometimes forget just how much rendering and state change is handled on the server with systems like Drupal. 

The service worker is definitely most effective with AJAX and browser-side rendering, when it can essentially just cache JSON payloads and use those to dynamically re-render the DOM as needed. The HTML, CSS and other static assets are stored too, but I'm not sure if caching those alone would do enough to enable offline functionality. I suppose if you could just cache a bunch of pre-rendered HTML, that would achieve a similar effect, but you'd have to anticipate a lot of user behavior. Also I'm not sure just how it works if you're trying to intercept plain HTTP requests for new page loads, instead of AJAX. I don't think the service worker is really intended for that. 

I'll explore the available options, but probably the best course, as you say, is to wait for the API support that comes with D8 and plan an incremental rollout of offline functionality. But that would, again as you say, entail re-implementing a lot of server-side UI modules as browser-side JS that could update the DOM on AJAX calls. Compared to that, writing the service worker is actually quite trivial.

Like I said, I sometimes forget in my React-centric world that not all rendering happens in the browser, and I hope I haven't taken too much for granted. I'm certainly open to contributing to more aspects of the UI, but it will be a considerable task. 

Are you still planning a Monthly Hangout call next Wednesday? Perhaps that would be a good opportunity to discuss strategy etc.

**Comment by mstenta** [comment]
Great info! Yes monthly call is next week - would love to talk more about it!

I wonder if the "offline" app could essentially just be it's own thing entirely, and when you "toggle" it, you're basically just hiding all the Drupal stuff and then just showing a completely separate JS app as a replacement (which works offline). If so, maybe we could use that to implement a super-simple feature to start, and grow from there. The one feature that I hear over and over is a simple observation app for snapping photos, taking notes, etc while you're in the field (without wifi). You can already do that with Observation logs in farmOS, but you need wifi.  It might be a simple use-case to start with... and could even be done in an add-on experimental module right now in Drupal 7 (ie: not in the core farmOS repo, so it doesn't hold up D8 migration and could be managed separately - or rewritten into the D8 codebase when the time comes).

Some work is beginning on a native app for simple observations here: https://github.com/farmOS/farmOS-mobile-app - So perhaps something like that - but built entirely in JS.

Got my gears turning...

**Comment by jgaehring** [comment]
Awesome, I'm psyched to talk more. I'll work on familiarizing myself more with the project, the UI and the codebase in the meantime. 

I'm also in planning mode right now for a personal OSS project, the gist of which is to design some light-weight, highly modular, component-based UI elements that could be used by farmers for anything from crop planning to direct marketing. Not my first thought when I reached out, but looking at it now there could be some synergy there, if you're open to discussing.

One concern regarding toggling: I think it would be important to make sure that the user can toggle to offline mode even after they've lost connectivity. I can see a lot of use cases, perhaps even a majority, when a farmer goes out to the field with no intention of using the app, but finds something she needs to note once she gets out there. It would defeat the purpose if she had to go back to an area of connectivity to load the cache before going into offline mode. I think it should be feasible to run background tasks (ie, polling) with the service worker while it has connectivity, and without the farmer doing a thing, so the cache is already pre-loaded when the occasion arises. That's a core part of the functionality of a service worker. But it might take some careful coordination with the Drupal UI. So that's somemething more to give your gears a spin. ;)

**Comment by mstenta** [comment]
@jgaehring and I started a module to experiment with offline farmOS features: https://github.com/jgaehring/farm_offline

Here's a rough outline of the plan for anyone interested: https://github.com/jgaehring/farm_offline/issues/2

**Comment by mstenta** [comment]
I'm going to close this issue and we can add new feature requests to the new repo(s) that @jgaehring is working on moving forward:

https://github.com/farmOS/farmOS-native
https://github.com/farmOS/farmOS-client
